File: DuudlagaFlow/Utilities/WAVEncoder.swift
import Foundation

struct WAVEncoder {
    static func encode(samples: [Float], sampleRate: Int = 16000, channels: Int = 1, bitDepth: Int = 16) -> Data {
        let int16Samples = samples.map { sample -> Int16 in
            let clamped = max(-1.0, min(1.0, sample))
            return Int16(clamped * Float(Int16.max))
        }

        let dataSize = int16Samples.count * MemoryLayout<Int16>.size
        let headerSize = 44
        let fileSize = headerSize + dataSize

        var data = Data(capacity: fileSize)

        // RIFF header
        data.append(contentsOf: "RIFF".utf8)
        data.append(littleEndian: UInt32(fileSize - 8))
        data.append(contentsOf: "WAVE".utf8)

        // fmt subchunk
        data.append(contentsOf: "fmt ".utf8)
        data.append(littleEndian: UInt32(16)) // subchunk size
        data.append(littleEndian: UInt16(1))  // PCM format
        data.append(littleEndian: UInt16(channels))
        data.append(littleEndian: UInt32(sampleRate))
        let byteRate = sampleRate * channels * (bitDepth / 8)
        data.append(littleEndian: UInt32(byteRate))
        let blockAlign = channels * (bitDepth / 8)
        data.append(littleEndian: UInt16(blockAlign))
        data.append(littleEndian: UInt16(bitDepth))

        // data subchunk
        data.append(contentsOf: "data".utf8)
        data.append(littleEndian: UInt32(dataSize))

        for sample in int16Samples {
            data.append(littleEndian: sample)
        }

        return data
    }
}

private extension Data {
    mutating func append(littleEndian value: UInt32) {
        var v = value.littleEndian
        append(Data(bytes: &v, count: MemoryLayout<UInt32>.size))
    }

    mutating func append(littleEndian value: UInt16) {
        var v = value.littleEndian
        append(Data(bytes: &v, count: MemoryLayout<UInt16>.size))
    }

    mutating func append(littleEndian value: Int16) {
        var v = value.littleEndian
        append(Data(bytes: &v, count: MemoryLayout<Int16>.size))
    }
}

File: windows/DuudlagaFlow/Utilities/WavEncoder.cs
using System.IO;

namespace DuudlagaFlow.Utilities;

public static class WavEncoder
{
    public static byte[] Encode(byte[] pcmData, int sampleRate = 16000, int channels = 1, int bitDepth = 16)
    {
        int dataSize = pcmData.Length;
        int fileSize = 44 + dataSize;

        using var ms = new MemoryStream(fileSize);
        using var writer = new BinaryWriter(ms);

        // RIFF header
        writer.Write("RIFF"u8);
        writer.Write(fileSize - 8);
        writer.Write("WAVE"u8);

        // fmt subchunk
        writer.Write("fmt "u8);
        writer.Write(16);                                    // subchunk size
        writer.Write((short)1);                              // PCM format
        writer.Write((short)channels);
        writer.Write(sampleRate);
        writer.Write(sampleRate * channels * bitDepth / 8);  // byte rate
        writer.Write((short)(channels * bitDepth / 8));      // block align
        writer.Write((short)bitDepth);

        // data subchunk
        writer.Write("data"u8);
        writer.Write(dataSize);
        writer.Write(pcmData);

        return ms.ToArray();
    }

    public static byte[] EncodeSilence(double durationSeconds, int sampleRate = 16000)
    {
        int sampleCount = (int)(sampleRate * durationSeconds);
        var pcm = new byte[sampleCount * 2]; // 16-bit = 2 bytes per sample
        // Fill with near-silence (tiny value to avoid API rejection)
        for (int i = 0; i < sampleCount; i++)
        {
            short value = 1;
            pcm[i * 2] = (byte)(value & 0xFF);
            pcm[i * 2 + 1] = (byte)(value >> 8);
        }
        return Encode(pcm, sampleRate);
    }
}
